File: samples/dbus.py
# simple script to show some info whenever properties chnaged
from gi.repository import GLib, Gio

def properties_changed_callback(
    connection,
    sender_name,
    object_path,
    interface_name,
    signal_name,
    parameters,
    user_data,
):
    print(
        f"sender: {sender_name}\n"
        f"object path: {object_path}\n"
        f"interface: {interface_name}\n"
        f"signal: {signal_name}"
    )

    default_props = connection.call_sync(
        "dev.cudiph.pulsedbus",
        object_path,
        "dev.cudiph.pulseaudio",
        "GetAllDefaultInformation",
        None,
        GLib.VariantType("(ssssdd)"),
        Gio.DBusCallFlags.NONE,
        -1,
        None,
    )

    _, _, sink_desc, source_desc, sink_vol, source_vol = default_props.unpack()
    print(
        f"{sink_desc} at {round(sink_vol * 100)}% volume\n"
        f"{source_desc} at {round(source_vol * 100)}% volume\n"
    )


main_loop = GLib.MainLoop()


bus = Gio.bus_get_sync(Gio.BusType.SESSION)

bus_id = bus.signal_subscribe(
    "dev.cudiph.pulsedbus",
    "org.freedesktop.DBus.Properties",
    "PropertiesChanged",
    "/dev/cudiph/pulsedbus",
    None,
    Gio.DBusSignalFlags.NONE,
    properties_changed_callback,
    None,
)


main_loop.run()

File: samples/dbus.lua
-- simple lua code to interact with pulsedbus service using dbusproxy class

local lgi = require("lgi")
local Gio, GLib = lgi.Gio, lgi.GLib


local main_loop = GLib.MainLoop(nil, false)

local proxy = Gio.DBusProxy.new_for_bus_sync(Gio.BusType.SESSION, Gio.DBusProxyFlags.NONE, nil, "dev.cudiph.pulsedbus",
    "/dev/cudiph/pulsedbus", "dev.cudiph.pulseaudio", nil, nil)


function proxy:on_g_signal(sender_name, signal_name, parameters, user_data)
    print(string.format("Signal received: %s", signal_name))
    local sink = self:get_cached_property("CurrentDefaultSink"):get_string()
    local sink_vol = self:get_cached_property("DefaultSinkVolume"):get_double() * 100
    sink_vol = sink_vol - math.floor(sink_vol) >= 0.5 and math.ceil(sink_vol) or math.floor(sink_vol)
    -- local source = self:get_cached_property("CurrentDefaultSource"):get_string()
    print(string.format("%s at %d%%\n", sink, sink_vol))
end

main_loop:run()
